hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java

```diff
@@ -107,25 +119,70 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
         // it must be submitted by an account with superuser privileges, that is checked during ingest
     }
 
+    public void handle(
+            @NonNull TransactionBody txn,
+            @NonNull final AdminServiceConfig adminServiceConfig,
+            @NonNull WritableSpecialFileStore specialFileStore,
+            @NonNull SwirldDualState dualState) {
+        final FreezeUpgradeActions upgradeActions =
+                new FreezeUpgradeActions(adminServiceConfig, dualState, specialFileStore);
+
+        requireNonNull(txn);
+        requireNonNull(adminServiceConfig);
+        requireNonNull(specialFileStore);
+        // TODO: requireNonNull(dualState);
+        // for the time being, this will always be null because we are not using SwirldDualState
+
+        FreezeTransactionBody freezeTxn = txn.freeze();
+        final Timestamp freezeStartTime = freezeTxn.startTimeOrThrow();
+        final Instant freezeStartTimeInstant =
+                Instant.ofEpochSecond(freezeStartTime.seconds(), freezeStartTime.nanos());
+        final FileID updateFileNum =
+                freezeTxn.updateFile(); // only some freeze types require this, it may be null for others
+
+        switch (freezeTxn.freezeType()) {
+            case PREPARE_UPGRADE -> {
+                requireNonNull(updateFileNum);
+                final Optional<byte[]> updateFileZip = specialFileStore.get(updateFileNum.fileNum());
+                if (updateFileZip.isEmpty()) throw new IllegalStateException("Update file not found");
```

Comment:
Is the *only* way this would be empty is if there is a bug in our software? Or can user mistakes cause it to happen? If a user mistake (say, a bad transaction) result in an empty file, then we must throw a HandleException with a proper response code.

Reply:
We've already checked this on preHandle so I don't know how it could revert to empty other than due to a bug in our software.